aspnet/mvc/controllers/testing.rst

```diff
@@ -1,8 +1,173 @@
-.. include:: /../common/stub-topic.txt
+Testing Controller Logic
+========================
+By `Steve Smith`_
 
-|stub-icon| Testing Controller Logic
-====================================
+Controllers in ASP.NET MVC apps should be small and focused on user-interface concerns. Large controllers that deal with non-UI concerns are more difficult to test and maintain.
 
-.. include:: /../common/stub-notice.txt
+.. contents:: Sections
+	:local:
+	:depth: 1
+	
+`View sample files <https://github.com/aspnet/Docs/tree/master/aspnet/mvc/controllers/testing/sample>`_
 
-.. _issue: https://github.com/aspnet/Docs/issues/123
+What is Controller Logic
+------------------------
+*Controllers* define groups of related *actions*. The grouping of related actions into controllers is useful for :doc:`routing </fundamentals/routing>`, applying :doc:`filters <filters>`, and :doc:`injecting common dependencies <dependency-injection>`. 
+
+:doc:`Learn more about controllers and actions <actions>`.
+
+Controller logic should be minimal and not be focused on business logic or infrastructure concerns (for example, data access). Test controller logic, not the framework. Test how the controller *behaves* based on valid or invalid inputs. Test controller responses based on the result of the business operation it performs.
+
+Typical controller responsibilities:
+	- Verify ``ModelState.IsValid``
+	- Return an error response if ``ModelState`` is invalid
+	- Retrieve a business entity from persistence
+	- Perform an action on the business entity
+	- Save the business entity to persistence
+	- Return an appropriate ``IActionResult``
+
+Unit Testing
```

Comment:
Most of the content in the Unit Testing section doesn't seem to document much that is specific to ASP.NET Core. This section demonstrates unit testing practices and related technologies (Moq, xUnit), but you could get the same information from a variety of existing sources. 

To make this section more relevant to ASP.NET Core and less of a unit testing demonstration I would suggest:
-  Add a deeper discussion of controller design concerns and the ASP.NET Core features that help you out (DI, options pattern, avoid statics, etc.)
- More discussion on the various things on Controller that you will probably need to mock (ActionContext, IUrlHelper, their various subproperties, etc.). Highlight that Controller is designed for unit testing.
- Reduce the amount of code that demonstrates general unit test writing practices. Focus on the stuff that is ASP.NET Core specific.